### PR TITLE
Make APIKeyMiddleware pure ASGI, so handlers can see a client disconnect

### LINKED ISSUE #208: APIKeyMiddleware is a BaseHTTPMiddleware, so no handler can see a client disconnect
`APIKeyMiddleware` (`server/auth.py`) is a Starlette `BaseHTTPMiddleware`, and
through it **`request.is_disconnected()` never reports a disconnect**. Any
handler that tries to notice a client leaving — and stop driving a device on
its behalf — silently can't.

## How it was found

PR #204 added `_run_until_client_leaves` to `server/api/device_ui.py`: run a long
device operation as a task, poll `request.is_disconnected()`, and cancel the work
and return 499 when the caller has gone. It exists because `scroll_to_element`
was measured sweeping for 413s and 523s against a client that had disconnected
at 180s — holding a simulator the next caller wanted, and swiping it while they
used it.

The guard's logic is correct and does not work in production. A review of #204
ran it on a scratch uvicorn behind each middleware in turn, disconnecting the
client after one second:

| Middleware stack | After the client disconnects |
|---|---|
| none | work cancelled, 499 |
| `CORSMiddleware` alone | work cancelled, 499 |
| `TimelineMiddleware` alone | work cancelled, 499 |
| **`APIKeyMiddleware` alone** | **no cancellation; ran to the end** |
| full production stack | no cancellation; ran to the end |

The auth middleware alone is sufficient to break it, and it is installed
unconditionally (`server/main.py:570`).

## Why rewriting the middleware, rather than working around it

A local workaround exists and was confirmed to work through the full stack:
watch `request.receive()` for `http.disconnect` in a background task and race it
against the work. It was deliberately *not* taken in #204, because:

- it relies on how `BaseHTTPMiddleware` wraps `receive` — the same class that
  just broke the obvious approach — so it builds on the quirk instead of removing
  it, and a Starlette upgrade can change it again;
- it has to be applied by hand to every long-running endpoint. #204 already found
  that guarding `scroll_to_element` alone left `tap_element` — which runs the same
  sweep by default — unguarded.

Rewriting `APIKeyMiddleware` as pure ASGI fixes disconnect detection for every
endpoint at once, including ones not yet written. #204 keeps its guard, which
starts working unchanged the moment this lands.

## Why it is smaller than "rewrite auth" sounds

- `server/auth.py` is 47 lines.
- `TimelineMiddleware` (`server/screenshot_timeline.py`) is already pure ASGI in
  this codebase — a working pattern to follow.
- `tests/conformance/test_auth.py` is a ready-made regression gate: it calls
  every protected operation (122 of them) with a missing token, an invalid
  bearer token and an invalid `X-API-Key`, checks every documented public path
  stays reachable, and includes a valid-key control so a server rejecting
  everything cannot pass.

## What to get right

It is the auth layer, so a mistake is a security hole rather than a scroll bug.
The realistic risks are mechanical, and each is a place this could go wrong
quietly:

- ASGI headers are a list of **lowercase byte** pairs, not a case-insensitive
  mapping — `authorization` / `x-api-key`, decoded.
- `scope["path"]` is not `request.url.path` under a mount or `root_path`; the
  public-path check must compare what the current code compares.
- the 401 must be hand-built as `http.response.start` + `http.response.body`,
  with the same JSON body the current `JSONResponse` sends.
- non-`http` scopes (`lifespan`, `websocket`) must pass straight through.
- the empty-key case: `server/auth.py` compares `X-API-Key` by equality against
  `""` as the missing-header default, which is only safe while the configured key
  is non-empty. Worth an explicit guard in the rewrite rather than an inherited
  assumption.

## When it lands

- run `tests/conformance/test_auth.py` against a server built from it — all
  schemes, all 122 operations — before and after, and require identical results;
- reproduce the disconnect table above through the full stack and see the 499;
- restore the 499 to `docs/api-reference.md` and the `scroll_to_element` MCP
  description, which #204 deliberately stopped advertising.

### PR DESCRIPTION
**Pull request by jerimiah797**
Fixes #208.

## The problem

`APIKeyMiddleware` was a Starlette `BaseHTTPMiddleware`, which replaces the receive channel with one that never delivers `http.disconnect` to the endpoint. So `request.is_disconnected()` answered False however long ago the caller had gone — and the middleware is installed unconditionally, so **no** handler could notice a disconnect.

That's why `scroll_to_element` was measured sweeping a simulator for 413s and 523s on behalf of a client that left at 180s (#84), holding a device the next caller wanted and swiping it while they used it.

## The change

`server/auth.py` is now pure ASGI, following `TimelineMiddleware`. The mechanical parts, each a place this could have gone quietly wrong:

- ASGI header names are lowercase **bytes** and values are undecoded; the first occurrence of a repeated header wins, as `Headers.get` does.
- The public list is matched against `scope["path"]`, which is what `request.url.path` read.
- The 401 is hand-built as `http.response.start` + `http.response.body`, byte-for-byte what `JSONResponse` sent, `content-length` included.
- `lifespan` and `websocket` scopes pass straight through.

Two deliberate differences:
- **An empty configured key now authorises nobody.** The old code compared `X-API-Key` against it with `""` as the default for a missing header, so an empty key accepted every request that sent no header at all. `ServerConfig` always generates a key, so this guards a state that shouldn't arise — which is what the old code assumed too.
- Keys are compared with `hmac.compare_digest`.

## Verification

**The tests were written against the old middleware first**, so the rewrite had to reproduce it rather than be described after the fact. 29 of 31 passed before the change; the two failures were the empty-key guard and the disconnect.

**Conformance auth suite** (`tests/conformance/test_auth.py` — every operation in the schema against a missing token, a wrong bearer token and a wrong `X-API-Key`, plus a valid-key control): **10 passed identically** against a server built from this branch and one built from `main`'s middleware.

**The disconnect, end to end.** #204's guard isn't merged, so I built a throwaway tree with both changes, ran the server, and aborted the same `scroll_to_element` after 4s:

| Middleware | After the client leaves |
|---|---|
| `main` (`BaseHTTPMiddleware`) | 35 tree reads, sweep ran to completion |
| this branch | 11 reads, `client disconnected — cancelling`, sweep stopped after 3 swipes |

**Mutants:** 16 of the middleware, all killed. Two survived the first pass and are worth naming: a `startswith` comparison that accepted any prefix of the key, and taking the last of two repeated headers instead of the first.

**Unit suite:** 2618 passed.

## Follow-up

Restoring the 499 to `docs/api-reference.md` and the `scroll_to_element` MCP description belongs on #204, which is where the guard and those docs live. `main` has no 499 to document. I'll do it there once this lands.

## Summary by CodeRabbit

- **Security**
  - Strengthened API-key validation and authentication for malformed, empty, non-ASCII, and missing keys.
  - Preserved access to approved public endpoints and supported authentication headers.

- **Bug Fixes**
  - Fixed client-disconnect detection after request bodies are read.
  - Long-running element interactions now stop device activity when callers disconnect and return HTTP 499.
  - Invalid API-key configuration now produces a clear startup error before server cleanup begins.

- **Documentation**
  - Documented disconnect behavior for element interaction tools.

### CODE REVIEW

**Comment by jerimiah797** [comment]
## Review of 13b0249

Code review, on my own worktree; nothing touched `~/.quern`, the server on 9100 or any device. What I did:
- read #208 and both versions of `server/auth.py`;
- exercised the middleware directly over ~30 key and header shapes;
- ran a scratch uvicorn on port 9199 with a throwaway key, for the 401 shape, HEAD, OPTIONS, CORS and live disconnects;
- ran 9 mutants against a `git archive` copy (tests: 37 passed).

**The rewrite is right in its essentials** — path semantics, header parsing, the byte-identical 401, non-http scopes, and the disconnect fix, which I reproduced live. **One finding is an auth bypass and should block merge.** Given this is the only review the PR gets, I'd want a second pair of eyes on whatever fix lands for it.

### High — auth bypass

**B-1. `encode("latin-1", errors="ignore")` silently shortens the key, and can empty it.** (Verified, two ways.)
`server/auth.py`, `_has_valid_key`: `expected = self.api_key.encode("latin-1", errors="ignore")`. Dropping the characters it can't encode means the key that is checked is not the key that was configured:

| configured key | request header | result |
|---|---|---|
| `abc🔑def` | `X-API-Key: abcdef` | **200** — a shorter key authenticates |
| `abc🔑def` | `X-API-Key: abc🔑def` (correct, UTF-8) | 401 — the real key is refused |
| `🔑🔑` | `Authorization: Bearer ` (empty token) | **200** — full bypass |

The last row is the bad one: with a key that has no latin-1-encodable characters, `expected` is `b""`, and `Bearer ` with nothing after it compares equal. That's a header anyone can send, with no knowledge of anything.

Reachable because the key is read from `~/.quern/api-key`, a user-editable file (`server/config.py:52-62`): `ServerConfig` generates a key only when that file is missing or empty, and otherwise uses its stripped contents as-is. Anyone who pastes a key containing a non-ASCII character — a smart quote, an emoji, a Cyrillic letter, a mangled copy-paste — gets this. The old code failed *closed* in both cases: it compared `str` to `str`, so a non-ASCII key refused everyone, including the legitimate client.

This is the same class as the empty-key case the PR already guards, one step along: "the configured key is not what gets compared".

Suggested fix — encode as UTF-8, which is what a client actually puts on the wire, and refuse an empty result:

```python
expected = self.api_key.encode("utf-8")
if not expected:
    return False
```

That also makes a non-ASCII key *work* rather than lock its owner out. Rejecting a non-ASCII key at config load would be a reasonable belt-and-braces addition. Either way, a test with a non-ASCII configured key belongs in the suite — mutant A3 (`errors="ignore"` → strict) survives today, which is how this stayed invisible.

### Medium — the disconnect claim is not quite "every endpoint"

**B-2. `TimelineMiddleware` re-breaks disconnects for its action endpoints while a timeline is active.** (Verified live.)
`server/screenshot_timeline.py:175`: when a timeline is active and the path is in `ACTION_FORMATTERS`, the request body is buffered and `receive` is replaced by

```python
async def cached_receive():
    return {"type": "http.request", "body": body_bytes}
```

which answers `http.request` forever and never delivers `http.disconnect` — the same shape of bug as `BaseHTTPMiddleware`. On a scratch stack of CORS → auth → timeline, with a timeline active and the client aborting after 1s:

- `/api/v1/device/ui/scroll-to-element` (not an action path): disconnect seen after 0.75s;
- `/api/v1/device/ui/tap-element` (an action path): **ran to completion, disconnect never seen**.

`tap-element` is the endpoint #204 added its guard for, so the gap lands on the case that motivated all of this — but only while a screenshot timeline is recording. Not a defect in this PR, and I wouldn't hold the merge for it: the fix is to yield the buffered body once and then delegate to the real `receive`. Worth a follow-up issue, and worth qualifying "fixes disconnect detection for every endpoint at once" in #208 and the CHANGELOG.

### Low — test gaps (each verified by a surviving mutant)
- **A9:** adding an OPTIONS exemption to the path check (`if scope.get("method") == "OPTIONS" or ...`) survives every test. In production CORS answers preflights above auth, so it is latent — but a blanket method exemption is exactly the change that should fail loudly. A test for OPTIONS to a protected path with no key → 401 would pin it.
- **A4:** `startswith(b"Bearer ")` → `startswith(b"Bearer")` survives. Not exploitable as far as I can tell, but `Authorization: Bearerabc` is untested.
- **A5:** `scope.get("headers") or ()` → `scope["headers"]` survives; the defensive `.get` is untested.
- **A2:** `hmac.compare_digest` → `==` survives, and can't be caught by a behavioural test. Worth a line of comment in the code so a later refactor doesn't quietly drop the constant-time compare.

### Loose ends
- #208's landing checklist says to restore the 499 to `docs/api-reference.md` and the `scroll_to_element` MCP description. Neither mentions 499 on this branch.
- I couldn't find `tests/conformance/test_auth.py` anywhere in the tree, so the "122 operations, before and after" gate can't be run from this branch. If it lives elsewhere, it's worth running before merge, and worth a pointer from the issue.

### Verified correct (no action)
- **Paths:** on Starlette 1.6, `URL(scope=scope).path` is `scope["path"]` with no `root_path` prefix, so swapping `request.url.path` for `scope["path"]` changes nothing, including under a mount. `PUBLIC_PATHS` matches main's tuple exactly, item for item.
- **Path shapes, all 401 (unchanged from main):** `/health/`, `//health`, `/HEALTH`, `/docs/oauth2-redirect`, `/health/secrets`, and a scope with no `path` key at all. Public paths still 200. Percent-encoding can't help an attacker here: uvicorn decodes before the scope, and a decoded path has to *equal* a public entry, which only public routes do.
- **Headers:** first occurrence wins for both headers, matching `Headers.get`; a good `X-API-Key` still wins after a bad `Authorization`; scheme matching stays case-sensitive; `compare_digest` on different-length bytes returns False without raising.
- **The 401:** byte-identical to what `JSONResponse` sent (same `separators`, `content-type: application/json`, `content-length: 39`), single body message, no `more_body`. Live: CORS headers are still applied to it, a preflight with `Origin` + `Access-Control-Request-Method` is answered by CORS above auth, `HEAD` gets 401 with no body, and keep-alive survives a 401 whose request body was never read (POST 401 then GET 200 on one connection).
- **The disconnect fix itself:** confirmed live through CORS → auth on a scratch server (disconnect seen 0.75s after a 1s abort). The reasoning generalises: the middleware now passes `receive` through untouched, so anything that respects it works, subject to B-2.
- **Deliberate changes:** the empty-key guard is right, and `compare_digest` is the correct call.

**Comment by jerimiah797** [comment]
Thank you — B-1 is a real bypass and it was mine. Fixed, with both of your reproductions as tests.

- **B-1 (high):** `latin-1` with `errors="ignore"` dropped unencodable characters from the configured key before comparing. Now a UTF-8 encode with an empty-result guard, so a non-ASCII key works rather than locking its owner out. Tests cover `abc<emoji>def` accepting `abcdef` (401 now), the real key (200), and an all-emoji key with an empty bearer token (401). Mutating the encode back kills a test.
- **B-2 (medium):** fixed here rather than deferred, since it re-broke the exact endpoint the guard was written for. `cached_receive` now delivers the buffered body once and delegates to the real `receive`. New `tests/test_timeline_middleware.py` covers the body still arriving and a disconnect reaching the endpoint. Live, with a timeline recording and `tap_element` aborted after 4s: `client disconnected — cancelling`, sweep stopped after 3 swipes.
- **Low:** `Bearer` without its space, a scope with no headers, and a method exemption are all pinned now. The `compare_digest` line carries a comment saying no test can distinguish it from `==`.

23 mutants across both middlewares, all killed. Unit suite: 2634.

On your two loose ends:
- `tests/conformance/test_auth.py` is on the `test/api-conformance-suite` branch (worktree `/Volumes/Home/jham/Dev/quern-conformance`), not on `main` — that's why you couldn't find it. It runs against a live server, so it's a live gate rather than something this branch can carry.
- The 499 docs belong on #204, where the guard and those docs live; `main` has no 499 to document. Noted on that PR.

Agreed on a second pair of eyes for the bypass fix specifically — flagging that to the user now.

**Comment by jerimiah797** [comment]
## Second review (independent reviewer, in its own worktree)

The bypass fix was reviewed adversarially by a second agent, since the first version of this code introduced a hole its own tests missed.

**Verdict: the fix holds, and would not block merge.** It attacked `_has_valid_key` at the wire level against a real uvicorn with a throwaway key `thr0waway-🔑-key` on ports 9251/9252, under **both** HTTP parsers:

| request | httptools | h11 |
|---|---|---|
| exact UTF-8 key, either header | 200 | 200 |
| the old latin-1-lossy form of the key | 401 | 401 |
| empty bearer token | 401 | 401 |
| obs-folded key | 400 | 401 |
| repeated header, wrong first | 401 | 401 |

It also confirmed the new tests **fail against the bypassed code** (reverting the encode is killed by 2 tests; removing the empty guard by 1), and that `_status` is a real 401 oracle rather than a recorder.

### Acted on

- **F1 (medium, the one real regression):** `is_disconnected()` reads the channel in an already-cancelled scope and discards non-disconnect messages. My replay returned the buffered body with no `await`, so the cancellation had nowhere to land: the body was consumed and discarded, and the endpoint's later `body()` waited on the real channel until the client left. A handler polling before reading its body hung until the disconnect it was polling for. The replay now yields first. Two tests, and removing the checkpoint kills one.
- **F2:** my "a non-ASCII key now works" claim was wrong — httpx and node both refuse to send one, so it was a silent lockout rather than a capability. `~/.quern/api-key` with non-ASCII content is now refused at startup, naming the file. Changelog corrected.
- **F6:** the key is encoded once in `__init__`, so an unencodable key fails at startup rather than per request.
- **F3:** the trailing-space assertion is re-scoped to the middleware's own rule. On the wire it was three mechanisms behind one claim (httptools strips leading not trailing, h11 strips both, TestClient strips neither).
- **F4, F5:** both surviving mutants are pinned — a path-less scope reading as public, and a present `X-API-Key` short-circuiting a good bearer token.
- **Minor:** the ASGI signatures now carry their Starlette types instead of `# noqa: ANN001`.

29 mutants across the two middlewares and the config loader, all killed. Unit suite: 2641.

**Comment by jerimiah797** [comment]
## Live verification (server built from this branch merged with #204)

**Auth matrix, by curl against a real server:**

| request | result |
|---|---|
| valid key, `Authorization: Bearer` / `X-API-Key` / lowercase `x-api-key` | 200 |
| wrong key, empty bearer token, first 20 chars of the key, no key | 401 |
| `/health`, `/tools`, `/openapi.json` | 200 without a key |
| `/health/secrets` | 401 (public paths match exactly) |
| 401 body | `content-type: application/json`, `content-length: 39`, `{"detail":"Invalid or missing API key"}` |
| CORS on the 401 | `access-control-allow-origin` present |
| OPTIONS preflight without a key | 200, answered above auth |

**Conformance auth gate:** 10 passed (every operation in the schema × missing / wrong-bearer / wrong-`X-API-Key`, plus the valid-key control).

**Full conformance suite:** 105 passed, 6 skipped, 1 failed — the failure is the known mock-404 bug (#182), unrelated.

**Disconnects, all four combinations** (request aborted after 4s; a full sweep is ~30 device reads):

| | cancelled | device reads |
|---|---|---|
| `scroll_to_element`, timeline idle | yes | 6 |
| `tap_element`, timeline idle | yes | 7 |
| `scroll_to_element`, timeline recording | yes | 2 |
| `tap_element`, timeline recording | yes | 6 |

Control: the same request allowed to finish returns 200 in 15.7s, so nothing is over-cancelling.

**Timeline still captures:** 73 entries recorded, every one with a screenshot, actions labelled correctly. That was the risk in changing how the body is replayed.

**MCP wrapper:** `list_devices` through `mcp/dist/index.js` reaches the server and is authorised — the one client path the unit tests can't exercise, since it builds the header itself.

**The new startup guard**, with `QUERN_STATE_DIR` pointed at a scratch dir holding a key with a smart quote: refuses to start, names the file, exit code 1, no traceback. An ASCII key in the same dir serves normally. The real `~/.quern/api-key` was untouched throughout.

One change came out of this: the guard originally surfaced as a raw traceback, which buried the sentence saying what to do. It now prints the reason and exits 1, via `_config_or_exit` so a test can drive it (2 mutants, both killed).

Unit suite: 2643.

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review

File: server/config.py [hunk 1]
```diff
@@ -55,6 +55,20 @@ def _load_or_create_api_key() -> str:
...
+                    # naming a character rather than the file it came from.
+                    # The server would start and refuse every request. This is
+                    # a hand-edited file, so a pasted smart quote is the likely
+                    # cause -- say that here rather than let it present as
+                    # "authentication is broken".
+                    raise ValueError(
+                        f"The API key in {API_KEY_FILE} contains non-ASCII "
+                        "characters, which HTTP headers cannot carry, so no "
+                        "client could authenticate. Edit the file, or delete "
+                        "it to have a new key generated."
```
[the remark this answers is not in the document]
Comment (jerimiah797) [reply server/config.py:87]: Confirmed and fixed in 262b0bc.

Measured rather than assumed, because httpx does not reject these where you might expect — `httpx.Headers({"Authorization": "Bearer abc\rdef"})` is built happily. It fails at the wire instead, against a running server:

```
CR:  LocalProtocolError: Illegal header value b'Bearer abc\rdef'
LF:  LocalProtocolError: Illegal header value b'Bearer abc\ndef'
NUL: LocalProtocolError: Illegal header value b'Bearer abc\x00def'
```

So the request is never sent, and the error names the character rather than the file it came from — the same failure the non-ASCII check above exists to prevent, one class down. Same refusal now, naming the file and the byte.

Rejecting `ord(c) < 0x20 or ord(c) == 0x7f`, after `.strip()`, so the trailing newline every editor adds is still stripped rather than refused. Four parametrised cases (CR, LF, NUL, DEL) plus one asserting surrounding whitespace is still accepted; all four fail with the check removed.

File: server/main.py [hunk 2]
```diff
@@ -864,10 +878,8 @@ def _bg_update_check():
 
         threading.Thread(target=_bg_update_check, daemon=True).start()
 
-    config = ServerConfig(
-        host=args.host,
-        port=server_port,
-        ring_buffer_size=args.buffer_size,
+    config = _config_or_exit(
+        host=args.host, port=server_port, ring_buffer_size=args.buffer_size,
     )
```
[the remark this answers is not in the document]
Comment (jerimiah797) [reply server/main.py:897]: Fixed in 262b0bc — and this was the consequential one.

Traced the order: the healthy-instance early exit, then `remove_state()`, then `reclaim_port()` (which kills stale quern processes), then `daemonize()`, and only then `ServerConfig`. Two consequences, both of which you named:

- An unusable key file destroyed the previous instance's leftovers for a start that never happened.
- `daemonize()` never returns in the parent, so past that point the message went to the child's log. The shell saw `quern start` exit 0 with no server running, which is worse than the error.

`_config_or_exit` is now called immediately after the early exit and before any of it. The later construction stays, with a comment saying why: the port is not resolved at the first call and nothing being validated depends on it, so building it twice is cheaper than threading a half-built config through port resolution.

Two tests: one asserts `remove_state`, `reclaim_port` and `daemonize` are never reached when the key file is unusable (it fails with the early call removed), and one asserts the "Server already running" path still comes first, so a broken key file on a machine with a healthy server does not turn that into a config complaint.

File: tests/test_config.py [hunk 3]
```diff
@@ -51,3 +51,64 @@ def test_channel_to_release_branch():
...
+
+        assert exit_info.value.code == 1
+        assert "api-key is unusable" in capsys.readouterr().out
+
+    def test_a_usable_key_file_is_returned(self, tmp_path, monkeypatch):
+        from server import main as main_mod
+
+        key_file = tmp_path / "api-key"
+        key_file.write_text("usable-key")
+        monkeypatch.setattr("server.config.API_KEY_FILE", key_file)
```
[the remark this answers is not in the document]
Comment (jerimiah797) [reply tests/test_config.py:148]: Fixed in 262b0bc. `CONFIG_DIR` is patched to `tmp_path` alongside `API_KEY_FILE` — loading creates the directory before it reads the key, so the test was making `~/.quern` on any machine that had none. That is the "tests never touch the real machine" rule in CONTRIBUTING, and the sibling tests in the same file were already doing it.

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review

**Comment by jerimiah797** [comment]
@coderabbitai review
